source/host_resolver.c

```diff
@@ -547,6 +567,61 @@ static inline void process_records(
     }
 }
 
+static void s_purge_host_cache_callback_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct host_purge_callback_options *options = arg;
+    aws_mem_release(options->allocator, task);
+    options->on_host_purge_complete(options->on_host_purge_complete_user_data);
+    aws_mem_release(options->allocator, options);
+}
+
+static int resolver_purge_host_cache(const struct aws_host_resolver_purge_host_options *options) {
+    struct default_host_resolver *default_host_resolver = options->resolver->impl;
+
+    AWS_LOGF_INFO(AWS_LS_IO_DNS, "id=%p: purging record for %s", (void *)options->resolver, options->host->bytes);
+
+    aws_mutex_lock(&default_host_resolver->resolver_lock);
+
+    struct aws_hash_element *element = NULL;
+    aws_hash_table_find(&default_host_resolver->host_entry_table, options->host, &element);
+
+    /* Success if entry doesn't exist in cache. */
+    if (element == NULL) {
+        aws_mutex_unlock(&default_host_resolver->resolver_lock);
+        if (options->on_host_purge_complete_callback != NULL) {
+            /*Sechedule completion callback asynchronouly */
+            struct host_purge_callback_options *purge_callback_options =
+                aws_mem_calloc(default_host_resolver->allocator, 1, sizeof(struct host_purge_callback_options));
+            purge_callback_options->allocator = default_host_resolver->allocator;
+            purge_callback_options->on_host_purge_complete = options->on_host_purge_complete_callback;
+            purge_callback_options->on_host_purge_complete_user_data = options->user_data;
```

Comment:
Should we keep the resolver alive until callback has fired? The callback doesn’t use resolver, so it is fine. However, it is possible that host_resolver destroy callback is fired before purge is complete.